app/admin.js

```diff
@@ -334,15 +484,282 @@ const handleVerification = co(function *(args) {
+    const filename = input.outputprefix + '.json';
+    assertFileDoesNotExist(filename);
+    const getPassword = function(i, n) {
+        if (i === n) {
+            return;
+        }
+        const passwordName = 'password' + i;
+        return input.getPassword(passwordName, 'Password for share ' + i + ': ', true)()
+            .then(function() {
+                return getPassword(i + 1, n);
+            });
+    };
+
+    return Q().then(function() {
+        console.log('Generate Split Keys');
+        console.log();
+    })
+        .then(input.getVariable('enter',"You are about to generate a key, press enter to continue\n"))
+        .then(input.getIntVariable('n', 'Number of shares per key (N): ', true, 1, 10))
+        .then(function() {
+            let mMin = 2;
+            if (input.n === 1) {
+                mMin = 1;
+            }
+            return input.getIntVariable('m', 'Number of shares required to restore key (M <= N): ', true, mMin, input.n)();
+        })
+        .then(input.getVariable('entropy', 'User supplied entropy string (optional): '))
+        .then(function() {
+            if (input.entropy) {
+                utils.addUserEntropy(input.entropy);
+            }
+            console.log("\n\nYou will now enter a password for each key share. These passwords must all be unique.");
+            console.log("\n\nAlso note, the order in which you enter these passwords will not matter when decrypting the file.");
+            return getPassword(0, input.n);
+
+        })
+        .then(function() {
+            const keys = _.range(0, 1).map(function(index) {
+                const key = utils.genSplitKey(input);
+                if (index % 10 === 0) {
+                    console.log('Generating key ' + index);
+                }
+                return {
+                    index: index,
+                    xpub: key.xpub,
+                    m: key.m,
+                    n: key.n,
+                    seedShares: key.seedShares
+                };
+            });
+            fs.writeFileSync(filename, JSON.stringify(keys, null, 2));
+            console.log('Wrote ' + filename);
+            const csvRows = keys.map(function(key) {
+                return key.index + ',' + key.xpub;
+            });
+        });
+};
+
+const handleRecoverKeys = co(function *(args) {
+   console.log('\n\nWARNING!! This will print a private key to the console. Continue at your own risk, or CTRL-C to exit\n');
+    const printKey = function(recoveredKeys) {
+        console.log(recoveredKeys);
+    }
+    decryptShardedKey(args, printKey);
+});
+
+
+/**
+ * Decrypt a key from the JSON file produced by initkey
+ * Note: by default, initkey only produces one key, so this will usually recover a list containing 1 key.
+ *
+ * This code has been refactored from a function that orignally decrypted a file containing multiple keys
+ *
+ */
+const decryptShardedKey = function(args, onFinish) {
```

Comment:
This function used to be handleRecoverKeys - but I changed the name because it is now being used by multiple other functions. Any time we must decrypt a sharded key file, it must go through this function. So it is used by Sign, Generate, and Recover